marketdata/charts.go

```diff
@@ -28,14 +29,34 @@ func InitCharts() *Charts {
 
 func (c *Charts) GetChartData(coin uint, token string, currency string, timeStart int64) (blockatlas.ChartData, error) {
 	chartsData := blockatlas.ChartData{}
-	for _, c := range c.chartProviders {
-		charts, err := c.GetChartData(coin, token, currency, timeStart)
-		if err != nil {
-			continue
-		}
-		return charts, nil
+	chartsDataChan := make(chan blockatlas.ChartData, len(c.chartProviders))
+	errChan := make(chan struct{})
+
+	var errCount int32
+	for _, provider := range c.chartProviders {
+		go func(provider chart.Provider) {
+			charts, err := provider.GetChartData(coin, token, currency, timeStart)
+			if err != nil || len(charts.Prices) == 0 {
+				if int(atomic.LoadInt32(&errCount)) == len(c.chartProviders) - 1 {
```

Comment:
why do you use `len(providers) - 1`? It means if you have 5 providers and 4 of them fail you will send error to err chan then it will be closed and if 5th provider also fails you will send error into closed errors channel and app will be crashed.

Reply:
if we have 5 providers, and at this iteration errCount == 4 (len(providers) - 1), it means that this provider is the last (5) and value will send to the errChan